src/backend/graphics/egl/wayland.rs

```diff
@@ -0,0 +1,322 @@
+                x if x == ffi::egl::TEXTURE_RGBA as i32 => Format::RGBA,
+                ffi::egl::TEXTURE_EXTERNAL_WL => Format::External,
+                ffi::egl::TEXTURE_Y_UV_WL => Format::Y_UV,
+                ffi::egl::TEXTURE_Y_U_V_WL => Format::Y_U_V,
+                ffi::egl::TEXTURE_Y_XUXV_WL => Format::Y_XUXV,
+                _ => panic!("EGL returned invalid texture type"),
+            };
+
+            let mut width: i32 = 0;
+            if unsafe { ffi::egl::QueryWaylandBufferWL(*display, buffer.ptr() as *mut _, ffi::egl::WIDTH as i32, &mut width as *mut _) == 0 } {
+                return Err(BufferAccessError::NotManaged(buffer));
+            }
+
+            let mut height: i32 = 0;
+            if unsafe { ffi::egl::QueryWaylandBufferWL(*display, buffer.ptr() as *mut _, ffi::egl::HEIGHT as i32, &mut height as *mut _) == 0 } {
+                return Err(BufferAccessError::NotManaged(buffer));
+            }
+
+            let mut inverted: i32 = 0;
+            if unsafe { ffi::egl::QueryWaylandBufferWL(*display, buffer.ptr() as *mut _, ffi::egl::WAYLAND_Y_INVERTED_WL, &mut inverted as *mut _) != 0 } {
+                inverted = 1;
+            }
+
+            let mut images = Vec::with_capacity(format.num_planes());
+            for i in 0..format.num_planes() {
+                let mut out = Vec::with_capacity(3);
+                out.push(ffi::egl::WAYLAND_PLANE_WL as i32);
+                out.push(i as i32);
+        		out.push(ffi::egl::NONE as i32);
+
+                images.push({
+        		    let image =
+                        unsafe { ffi::egl::CreateImageKHR(
+                            *display,
+                            ffi::egl::NO_CONTEXT,
+                            ffi::egl::WAYLAND_BUFFER_WL,
+            				buffer.ptr() as *mut _,
+            				out.as_ptr(),
+                        ) };
+                    if image == ffi::egl::NO_IMAGE_KHR {
+                        return Err(BufferAccessError::EGLImageCreationFailed);
+                    } else {
+                        image
+                    }
+                });
+            }
+
+            Ok(EGLImages {
+                display: Rc::downgrade(&display),
+                width: width as u32,
+                height: height as u32,
+                y_inverted: inverted != 0,
+                format,
+                images,
+                buffer,
+            })
+        } else {
+            Err(BufferAccessError::ContextLost)
+        }
+    }
+}
+
+impl<E: EGLWaylandExtensions> EGLWaylandExtensions for Rc<E>
+{
+    fn bind_wl_display(&self, display: &Display) -> Result<EGLDisplay> {
+        (**self).bind_wl_display(display)
+    }
+    fn unbind_wl_display(&self, display: &Display) -> Result<()> {
+        (**self).unbind_wl_display(display)
+    }
+}
+
+impl<B: native::Backend, N: native::NativeDisplay<B>> EGLWaylandExtensions for EGLContext<B, N> {
```

Comment:
I see you've added a lot of implementations of this traits to various types that embed an `EGLContext`, which seem redundant, what's the reason of that?

Reply:
I could make this a `pub(crate)` function of `EGLContext` instead, that every other implementation calls into. But I think this is unnecessary and just hides more options from the user. I just implemented the trait for every struct it made sense for.